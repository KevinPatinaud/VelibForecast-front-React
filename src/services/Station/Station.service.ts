import { getServerURL } from "../../helper/Utils";
import { Station } from "../../model/Station";
import { StationState } from "../../model/StationState";
import HttpService from "../Http/Http.service";

const getStations = async () => {
  const httpResponse = await HttpService.get(
    getServerURL() + ":8083/api/station/"
  );

  const stations = [];
  for (let i = 0; i < httpResponse.data.length; i++) {
    const station = {
      id: httpResponse.data[i]?.stationCode,
      name: httpResponse.data[i]?.name,
      lat: httpResponse.data[i]?.latitude,
      lng: httpResponse.data[i]?.longitude,
      state: {
        nmbBikeAvailable: httpResponse.data[i]?.states
          ? httpResponse.data[i]?.states?.numBikesAvailable
          : undefined,
        nmbPlaceAvailable: httpResponse.data[i]?.states
          ? httpResponse.data[i]?.states?.numDocksAvailable
          : undefined,
      },
    } as Station;
    stations.push(station);
  }
  return stations;
};

const getStatus = async () => {
  const httpResponse = await HttpService.get(
    getServerURL() + ":8083/api/station/states"
  );

  const states = [];
  for (let i = 0; i < httpResponse.data.length; i++) {
    const state = {
      idStation: httpResponse.data[i]?.station?.stationCode,
      nmbBikeAvailable: httpResponse.data[i]?.numBikesAvailable,
      nmbPlaceAvailable: httpResponse.data[i]?.numDocksAvailable,
    } as StationState;
    states.push(state);
  }
  return states;
};

const getStatusInFutur = async (idStation: Number, inMinutes: Number) => {
  const httpResponse = await HttpService.get(
    getServerURL() +
      ":8083/api/station/" +
      idStation +
      "/state?inMinutes=" +
      inMinutes
  );

  return {
    idStation: idStation,
    nmbBikeAvailable: httpResponse.data?.numBikesAvailable,
    nmbPlaceAvailable: httpResponse.data?.numDockAvailable,
  } as StationState;
};

const StationService = {
  getStations,
  getStatus,
  getStatusInFutur,
};

export default StationService;
