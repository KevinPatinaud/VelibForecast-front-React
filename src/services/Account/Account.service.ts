import { AxiosResponse } from "axios";
import { getServerURL } from "../../helper/Utils";
import { Account } from "../../model/Account";
import { Station } from "../../model/Station";
import HTTPService from "../Http/Http.service";
import HttpService from "../Http/Http.service";

const connectMailAccount = async (account: Account, captchaToken: String) => {
  const result = await HttpService.put(
    getServerURL() + ":8083/api/user/mailuser",
    {
      email: account.email,
      captchaToken: captchaToken,
    },
    {
      auth: {
        username: account.email,
        password: account.password,
      },
    }
  );

  return interpretConection(result);
};
const connectFacebookAccount = async (accessToken: String) => {
  HTTPService.setHeader("facebook_access_token", accessToken.toString());
  const result = await HttpService.put(
    getServerURL() + ":8083/api/user/facebookuser"
  );
  HTTPService.removeHeader("facebook_access_token");

  return interpretConection(result);
};

const createMailAccount = async (account: Account, captchaToken: String) => {
  const result = await HttpService.post(
    getServerURL() + ":8083/api/user/mailuser",
    {
      email: account.email,
      password: account.password,
      captchaToken: captchaToken,
    }
  );

  return interpretConection(result);
};

const createFacebookAccount = async (accessToken: String) => {
  const result = await HttpService.post(
    getServerURL() + ":8083/api/user/facebookuser",
    {
      accessToken: accessToken,
    }
  );

  return interpretConection(result);
};

const interpretConection = (result: AxiosResponse<any, any>) => {
  if (result === undefined || result.status === undefined) return null;

  if (result.status === 200) {
    const jwt = result?.data.JWT as String;
    HttpService.setAuthToken(jwt.toString());
    return getUserFromJWT(jwt.toString());
  } else return result.status;
};

const getUserFromJWT = (jwt: string) => {
  if (jwt.split(".").length !== 3) return {} as Account;

  const payload = JSON.parse(window.atob(jwt.split(".")[1]));

  const sub = JSON.parse(payload.sub);

  const user = {
    id: sub.iduser,
    favoriteStations: interpretJSONStationReceive(sub.favoriteStations),
  } as Account;
  return user;
};

const interpretJSONStationReceive = (favoriteStations: any) => {
  const stations = [] as Station[];

  favoriteStations.forEach(function (favStation: any) {
    stations.push({
      id: favStation.stationCode,
      name: favStation.name,
      capacity: favStation.capacity,
      lat: favStation.latitude,
      lng: favStation.longitude,
    } as Station);
  });

  return stations;
};

const isAccountExist = async (id: String) => {
  return (
    await HttpService.get(
      getServerURL() + ":8083/api/user/ismailalreadyrecorded?mail=" + id
    )
  )?.data;
};

const isAuthTokenSetted = () => {
  return HttpService.isAuthTokenSetted();
};

const disconnect = () => {
  HttpService.setAuthToken();
};

const addFavoriteStation = (station: Station) => {
  HTTPService.put(getServerURL() + ":8083/api/user/addfavoritestation", {
    id_station: station.id,
  });
};

const removeFavoriteStation = (station: Station) => {
  HTTPService.put(getServerURL() + ":8083/api/user/removefavoritestation", {
    id_station: station.id,
  });
};

const AccountService = {
  connectMailAccount,
  connectFacebookAccount,
  createMailAccount,
  createFacebookAccount,
  isAccountExist,
  isAuthTokenSetted,
  interpretConection,
  getUserFromJWT,
  disconnect,
  addFavoriteStation,
  removeFavoriteStation,
};

export default AccountService;
