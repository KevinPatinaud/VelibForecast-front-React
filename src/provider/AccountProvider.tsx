import React, { useMemo, useState } from "react";
import { FC } from "react";
import { Account } from "../model/Account";

export const AccountContext = React.createContext({
  account: {
    isConnected: false,
  } as Account,
  setAccount: (account: Account) => {},
});

const AccountProvider: FC<{ children: JSX.Element }> = ({ children }) => {
  const [account, setAccount] = useState(() => {
    const accountSession = sessionStorage.getItem("ACCOUNT");
    if (accountSession === null || accountSession === "")
      return {
        isConnected: false,
      } as Account;
    else return JSON.parse(accountSession) as Account;
  });

  const setSessionAccount = (account: Account) => {
    sessionStorage.setItem("ACCOUNT", JSON.stringify(account));
    setAccount(account);
  };

  const value = useMemo(
    () => ({ account: account, setAccount: setSessionAccount }),
    [account]
  );

  return (
    <AccountContext.Provider value={value}>{children}</AccountContext.Provider>
  );
};

export default AccountProvider;
