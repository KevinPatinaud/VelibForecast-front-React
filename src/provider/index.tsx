import { FC } from "react";
import AccountProvider from "./AccountProvider";
import ErrorBoundary from "./ErrorBoundary";
import IntlProvider from "./IntlProvider";

const AppProviders: FC<{ children: JSX.Element }> = ({ children }) => {
  return (
    <ErrorBoundary fallback={<></>}>
      <AccountProvider>
        <IntlProvider>{children}</IntlProvider>
      </AccountProvider>
    </ErrorBoundary>
  );
};

export default AppProviders;
