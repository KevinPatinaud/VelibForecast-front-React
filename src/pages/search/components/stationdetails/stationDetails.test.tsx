import { render } from "@testing-library/react";
import { Station } from "../../../../model/Station";
import StationDetails from "./stationDetails";
import wrapper from "../../../../helper/test-context-builder";
import StationService from "../../../../services/Station/Station.service";
import { StationState } from "../../../../model/StationState";
import { AccountContext } from "../../../../provider/AccountProvider";
import { Account } from "../../../../model/Account";
import userEvent from "@testing-library/user-event";
import AccountService from "../../../../services/Account/Account.service";

jest
  .spyOn(StationService, "getStatusInFutur")
  .mockResolvedValue({} as StationState);

describe("<StationDetails>", () => {
  describe("When is rendering", () => {
    it("should be displayed", () => {
      const scr = render(
        <AccountContext.Provider
          value={{
            account: {
              isConnected: true,
              favoriteStations: [
                {
                  id: 123,
                  name: "123",
                  lat: 12,
                  lng: 42,
                  capacity: 25,
                  state: {} as StationState,
                } as Station,
                {
                  id: 456,
                  name: "456",
                  lat: 12,
                  lng: 42,
                  capacity: 25,
                  state: {} as StationState,
                } as Station,
              ],
            } as Account,
            setAccount: jest.fn(),
          }}
        >
          <StationDetails
            station={
              {
                id: 123,
                name: "Benjamin Godard - Victor Hugo",
                lat: 48.865983,
                lng: 2.275725,
                capacity: 35,
              } as Station
            }
          />
        </AccountContext.Provider>,
        { wrapper }
      );

      expect(scr).toMatchSnapshot();
    });
  });

  describe("When the user click on the add to favorite button", () => {
    it("should make a api call", () => {
      jest
        .spyOn(AccountService, "addFavoriteStation")
        .mockImplementation(() => {});

      const scr = render(
        <AccountContext.Provider
          value={{
            account: {
              isConnected: true,
              favoriteStations: [
                {
                  id: 123,
                  name: "123",
                  lat: 12,
                  lng: 42,
                  capacity: 25,
                  state: {} as StationState,
                } as Station,
                {
                  id: 456,
                  name: "456",
                  lat: 12,
                  lng: 42,
                  capacity: 25,
                  state: {} as StationState,
                } as Station,
              ],
            } as Account,
            setAccount: jest.fn(),
          }}
        >
          <StationDetails
            station={
              {
                id: 789,
                name: "Benjamin Godard - Victor Hugo",
                lat: 48.865983,
                lng: 2.275725,
                capacity: 35,
              } as Station
            }
          />
        </AccountContext.Provider>,
        { wrapper }
      );

      userEvent.click(scr.getByTestId("add_fav_station"));

      expect(AccountService.addFavoriteStation).toHaveBeenCalledWith({
        capacity: 35,
        id: 789,
        lat: 48.865983,
        lng: 2.275725,
        name: "Benjamin Godard - Victor Hugo",
      });
    });
  });

  describe("When the user click on the remove favorite button", () => {
    it("should make an api call", () => {
      jest
        .spyOn(AccountService, "removeFavoriteStation")
        .mockImplementation(() => {});

      const scr = render(
        <AccountContext.Provider
          value={{
            account: {
              isConnected: true,
              favoriteStations: [
                {
                  id: 123,
                  name: "123",
                  lat: 12,
                  lng: 42,
                  capacity: 25,
                  state: {} as StationState,
                } as Station,
                {
                  id: 456,
                  name: "456",
                  lat: 12,
                  lng: 42,
                  capacity: 25,
                  state: {} as StationState,
                } as Station,
              ],
            } as Account,
            setAccount: jest.fn(),
          }}
        >
          <StationDetails
            station={
              {
                id: 123,
                name: "Benjamin Godard - Victor Hugo",
                lat: 48.865983,
                lng: 2.275725,
                capacity: 35,
              } as Station
            }
          />
        </AccountContext.Provider>,
        { wrapper }
      );

      userEvent.click(scr.getByTestId("remove_fav_station"));

      expect(AccountService.removeFavoriteStation).toHaveBeenCalledWith({
        capacity: 35,
        id: 123,
        lat: 48.865983,
        lng: 2.275725,
        name: "Benjamin Godard - Victor Hugo",
      });
    });
  });
});
