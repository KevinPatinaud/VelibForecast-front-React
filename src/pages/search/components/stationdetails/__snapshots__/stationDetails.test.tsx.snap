// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StationDetails> When is rendering should be displayed 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="titleStation"
        >
          <div
            class="stationName"
          >
            Benjamin Godard - Victor Hugo
          </div>
          <div
            class="removeFavoriteStation"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-heart-circle-minus "
              data-icon="heart-circle-minus"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 576 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M47.6 300.4L228.3 469.1c7.5 7 17.4 10.9 27.7 10.9s20.2-3.9 27.7-10.9l2.6-2.4C267.2 438.6 256 404.6 256 368c0-97.2 78.8-176 176-176c28.3 0 55 6.7 78.7 18.5c.9-6.5 1.3-13 1.3-19.6v-5.8c0-69.9-50.5-129.5-119.4-141C347 36.5 300.6 51.4 268 84L256 96 244 84c-32.6-32.6-79-47.5-124.6-39.9C50.5 55.6 0 115.2 0 185.1v5.8c0 41.5 17.2 81.2 47.6 109.5zM576 368c0-79.5-64.5-144-144-144s-144 64.5-144 144s64.5 144 144 144s144-64.5 144-144zm-64 0c0 8.8-7.2 16-16 16H368c-8.8 0-16-7.2-16-16s7.2-16 16-16H496c8.8 0 16 7.2 16 16z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
        <table
          class="detailTable"
        >
          <thead>
            <tr>
              <th />
              <th>
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-bicycle "
                  data-icon="bicycle"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 640 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M312 32c-13.3 0-24 10.7-24 24s10.7 24 24 24h25.7l34.6 64H222.9l-27.4-38C191 99.7 183.7 96 176 96H120c-13.3 0-24 10.7-24 24s10.7 24 24 24h43.7l22.1 30.7-26.6 53.1c-10-2.5-20.5-3.8-31.2-3.8C57.3 224 0 281.3 0 352s57.3 128 128 128c65.3 0 119.1-48.9 127-112h49c8.5 0 16.3-4.5 20.7-11.8l84.8-143.5 21.7 40.1C402.4 276.3 384 312 384 352c0 70.7 57.3 128 128 128s128-57.3 128-128s-57.3-128-128-128c-13.5 0-26.5 2.1-38.7 6L375.4 48.8C369.8 38.4 359 32 347.2 32H312zM458.6 303.7l32.3 59.7c6.3 11.7 20.9 16 32.5 9.7s16-20.9 9.7-32.5l-32.3-59.7c3.6-.6 7.4-.9 11.2-.9c39.8 0 72 32.2 72 72s-32.2 72-72 72s-72-32.2-72-72c0-18.6 7-35.5 18.6-48.3zM133.2 368h65c-7.3 32.1-36 56-70.2 56c-39.8 0-72-32.2-72-72s32.2-72 72-72c1.7 0 3.4 .1 5.1 .2l-24.2 48.5c-9 18.1 4.1 39.4 24.3 39.4zm33.7-48l50.7-101.3 72.9 101.2-.1 .1H166.8zm90.6-128H365.9L317 274.8 257.4 192z"
                    fill="currentColor"
                  />
                </svg>
                 
                vélo(s) disponnible(s)
              </th>
              <th>
                <b>
                  P
                </b>
                 
                place(s) disponnible(s)
              </th>
            </tr>
          </thead>
          <tbody>
            <tr
              class="highlightLine"
            >
              <td
                class="firstCol"
              >
                Actuellement
              </td>
              <td />
              <td />
            </tr>
            <tr>
              <td
                class="firstCol"
              >
                Dans 1 heure
              </td>
              <td />
              <td />
            </tr>
            <tr
              class="highlightLine"
            >
              <td
                class="firstCol"
              >
                Dans 2 heure
              </td>
              <td />
              <td />
            </tr>
            <tr>
              <td
                class="firstCol"
              >
                Dans 3 heure
              </td>
              <td />
              <td />
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="titleStation"
      >
        <div
          class="stationName"
        >
          Benjamin Godard - Victor Hugo
        </div>
        <div
          class="removeFavoriteStation"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-heart-circle-minus "
            data-icon="heart-circle-minus"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 576 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M47.6 300.4L228.3 469.1c7.5 7 17.4 10.9 27.7 10.9s20.2-3.9 27.7-10.9l2.6-2.4C267.2 438.6 256 404.6 256 368c0-97.2 78.8-176 176-176c28.3 0 55 6.7 78.7 18.5c.9-6.5 1.3-13 1.3-19.6v-5.8c0-69.9-50.5-129.5-119.4-141C347 36.5 300.6 51.4 268 84L256 96 244 84c-32.6-32.6-79-47.5-124.6-39.9C50.5 55.6 0 115.2 0 185.1v5.8c0 41.5 17.2 81.2 47.6 109.5zM576 368c0-79.5-64.5-144-144-144s-144 64.5-144 144s64.5 144 144 144s144-64.5 144-144zm-64 0c0 8.8-7.2 16-16 16H368c-8.8 0-16-7.2-16-16s7.2-16 16-16H496c8.8 0 16 7.2 16 16z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
      <table
        class="detailTable"
      >
        <thead>
          <tr>
            <th />
            <th>
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-bicycle "
                data-icon="bicycle"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 640 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M312 32c-13.3 0-24 10.7-24 24s10.7 24 24 24h25.7l34.6 64H222.9l-27.4-38C191 99.7 183.7 96 176 96H120c-13.3 0-24 10.7-24 24s10.7 24 24 24h43.7l22.1 30.7-26.6 53.1c-10-2.5-20.5-3.8-31.2-3.8C57.3 224 0 281.3 0 352s57.3 128 128 128c65.3 0 119.1-48.9 127-112h49c8.5 0 16.3-4.5 20.7-11.8l84.8-143.5 21.7 40.1C402.4 276.3 384 312 384 352c0 70.7 57.3 128 128 128s128-57.3 128-128s-57.3-128-128-128c-13.5 0-26.5 2.1-38.7 6L375.4 48.8C369.8 38.4 359 32 347.2 32H312zM458.6 303.7l32.3 59.7c6.3 11.7 20.9 16 32.5 9.7s16-20.9 9.7-32.5l-32.3-59.7c3.6-.6 7.4-.9 11.2-.9c39.8 0 72 32.2 72 72s-32.2 72-72 72s-72-32.2-72-72c0-18.6 7-35.5 18.6-48.3zM133.2 368h65c-7.3 32.1-36 56-70.2 56c-39.8 0-72-32.2-72-72s32.2-72 72-72c1.7 0 3.4 .1 5.1 .2l-24.2 48.5c-9 18.1 4.1 39.4 24.3 39.4zm33.7-48l50.7-101.3 72.9 101.2-.1 .1H166.8zm90.6-128H365.9L317 274.8 257.4 192z"
                  fill="currentColor"
                />
              </svg>
               
              vélo(s) disponnible(s)
            </th>
            <th>
              <b>
                P
              </b>
               
              place(s) disponnible(s)
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class="highlightLine"
          >
            <td
              class="firstCol"
            >
              Actuellement
            </td>
            <td />
            <td />
          </tr>
          <tr>
            <td
              class="firstCol"
            >
              Dans 1 heure
            </td>
            <td />
            <td />
          </tr>
          <tr
            class="highlightLine"
          >
            <td
              class="firstCol"
            >
              Dans 2 heure
            </td>
            <td />
            <td />
          </tr>
          <tr>
            <td
              class="firstCol"
            >
              Dans 3 heure
            </td>
            <td />
            <td />
          </tr>
        </tbody>
      </table>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
